gdx/src/com/badlogic/gdx/maps/tiled/BaseTmjMapLoader.java

```diff
@@ -0,0 +1,752 @@
+							throw new GdxRuntimeException("Error Reading TMJ Layer Data: Premature end of tile data");
+						ids[y * width + x] = unsignedByteToInt(temp[0]) | unsignedByteToInt(temp[1]) << 8
+							| unsignedByteToInt(temp[2]) << 16 | unsignedByteToInt(temp[3]) << 24;
+					}
+				}
+			} catch (IOException e) {
+				throw new GdxRuntimeException("Error Reading TMJ Layer Data - IOException: " + e.getMessage());
+			} finally {
+				StreamUtils.closeQuietly(is);
+			}
+		} else {
+			// any other value of 'encoding' is one we're not aware of, probably a feature of a future version of Tiled
+			// or another editor
+			throw new GdxRuntimeException("Unrecognised encoding (" + encoding + ") for TMJ Layer Data");
+		}
+
+		return ids;
+	}
+
+	private static int unsignedByteToInt (byte b) {
+		return b & 0xFF;
+	}
+
+	private static float[] convertArrayToPrimitive (List<Float> list) {
+		float[] ret = new float[list.size()];
+		Iterator<Float> iterator = list.iterator();
+		for (int i = 0; i < ret.length; i++) {
+			ret[i] = iterator.next().intValue();
+		}
+		return ret;
+	}
+
+	protected static FileHandle getRelativeFileHandle (FileHandle file, String path) {
+		StringTokenizer tokenizer = new StringTokenizer(path, "\\/");
+		FileHandle result = file.parent();
+		while (tokenizer.hasMoreElements()) {
+			String token = tokenizer.nextToken();
+			if (token.equals(".."))
+				result = result.parent();
+			else {
+				result = result.child(token);
+			}
+		}
+		return result;
+	}
+
+	@Override
+	public Array<AssetDescriptor> getDependencies (String fileName, FileHandle tmjFile, P parameter) {
+		this.root = json.parse(tmjFile);
+
+		TextureLoader.TextureParameter textureParameter = new TextureLoader.TextureParameter();
+		if (parameter != null) {
+			textureParameter.genMipMaps = parameter.generateMipMaps;
+			textureParameter.minFilter = parameter.textureMinFilter;
+			textureParameter.magFilter = parameter.textureMagFilter;
+		}
+
+		return getDependencyAssetDescriptors(tmjFile, textureParameter);
+	}
+
+	/** Gets a map of the object ids to the {@link MapObject} instances. Returns null if
+	 * {@link #loadTiledMap(FileHandle, BaseTmjMapLoader.Parameters, ImageResolver)} has not been called yet.
+	 *
+	 * @return the map of the ids to {@link MapObject}, or null if
+	 *         {@link #loadTiledMap(FileHandle, BaseTmjMapLoader.Parameters, ImageResolver)} method has not been called yet. */
+	public @Null IntMap<MapObject> getIdToObject () {
+		return idToObject;
+	}
+
+	protected abstract Array<AssetDescriptor> getDependencyAssetDescriptors (FileHandle tmjFile,
+		TextureLoader.TextureParameter textureParameter);
+
+	/** Loads the map data, given the JSON root element
+	 *
+	 * @param tmjFile the Filehandle of the tmj file
+	 * @param parameter
+	 * @param imageResolver
+	 * @return the {@link TiledMap} */
+	protected TiledMap loadTiledMap (FileHandle tmjFile, P parameter, ImageResolver imageResolver) {
+		this.map = new TiledMap();
```

Comment:
Oh yea, that's the big issue I had with this class as well. It IS literally a copy paste. But using the JSON parser instead of the XML one and there a small differences in the properties between the json format and the xml.
So should we keep it small and build out a utility class for some shared functionality?
Or do a larger refactor. I feel as though having BaseTmxMapLoader and BaseTmjMapLoader having a shared parent class would make sense? A BaseMapLoader class which they both could extend? But that seems like it's leaking outside of the original intent of the PR.
What do you think? Keep it small?

Reply:
I think keep it as it is for now. But should be refactored in the future at some point.

Reply:
Alright, Sounds like a plan! I actually started a refactor locally. But then realized if we made such sweeping changes right now it's going to have big consequences for all of the 8+ tiled PRs that are waiting to get merged. Once everything is stable we can worry about a large refactor.

Thanks for the input!

Reply:
Ok, so one last follow up. Need your opinion. 
So I'm building out a small utility class which should clean up a bunch of duplicated methods that are not specific to XML vs Json parsing.
```
public class TiledMapLoaderUtils {

	 protected static final int FLAG_FLIP_HORIZONTALLY = 0x80000000;
	 protected static final int FLAG_FLIP_VERTICALLY = 0x40000000;
	 protected static final int FLAG_FLIP_DIAGONALLY = 0x20000000;
	 protected static final int MASK_CLEAR = 0xE0000000;

	 public static Object castProperty (String name, String value, String type) {
		  if (type == null || "string".equals(type)) {
				return value;
		  } else if (type.equals("int")) {
				return Integer.valueOf(value);
		  } else if (type.equals("float")) {
				return Float.valueOf(value);
		  } else if (type.equals("bool")) {
				return Boolean.valueOf(value);
		  } else if (type.equals("color")) {
				// Tiled uses the format #AARRGGBB
				String opaqueColor = value.substring(3);
				String alpha = value.substring(1, 3);
				return Color.valueOf(opaqueColor + alpha);
		  } else {
				throw new GdxRuntimeException(
					"Wrong type given for property " + name + ", given : " + type + ", supported : string, bool, int, float, color");
		  }
	 }

	 public static Cell createTileLayerCell (boolean flipHorizontally, boolean flipVertically, boolean flipDiagonally) {
		  Cell cell = new Cell();
		  if (flipDiagonally) {
				if (flipHorizontally && flipVertically) {
					 cell.setFlipHorizontally(true);
					 cell.setRotation(Cell.ROTATE_270);
				} else if (flipHorizontally) {
					 cell.setRotation(Cell.ROTATE_270);
				} else if (flipVertically) {
					 cell.setRotation(Cell.ROTATE_90);
				} else {
					 cell.setFlipVertically(true);
					 cell.setRotation(Cell.ROTATE_270);
				}
		  } else {
				cell.setFlipHorizontally(flipHorizontally);
				cell.setFlipVertically(flipVertically);
		  }
		  return cell;
	 }

	 public static FileHandle getRelativeFileHandle(FileHandle file, String path) {
		  StringTokenizer tokenizer = new StringTokenizer(path, "\\/");
		  FileHandle result = file.parent();
		  while (tokenizer.hasMoreElements()) {
				String token = tokenizer.nextToken();
				if (token.equals(".."))
					 result = result.parent();
				else {
					 result = result.child(token);
				}
		  }
		  return result;
	 }

	 public static void addStaticTiledMapTile(TiledMapTileSet tileSet, TextureRegion textureRegion, int tileId, float offsetX, float offsetY, boolean flipY) {
		  TiledMapTile tile = new StaticTiledMapTile(textureRegion);
		  tile.setId(tileId);
		  tile.setOffsetX(offsetX);
		  tile.setOffsetY(flipY ? -offsetY : offsetY);
		  tileSet.putTile(tileId, tile);
	 }

	 public static int unsignedByteToInt (byte b) {
		  return b & 0xFF;
	 }

protected void addStaticTiledMapTile (TiledMapTileSet tileSet, TextureRegion textureRegion, int tileId, float offsetX,
		float offsetY) {
		 TiledMapLoaderUtils.addStaticTiledMapTile(tileSet, textureRegion, tileId, offsetX, offsetY, flipY);
	}
```

But now here's a question, almost all of these methods in the BaseTmxMapLoader were set as protected.
Examples: "protected Cell createTileLayerCell()","protected Object castProperty()","protected static FileHandle getRelativeFileHandle()","protected static int unsignedByteToInt()"

So proper convention would say we shouldn't get rid of these in the BaseTmxMapLoader right? Just in case there might be someone out there using this codebase with their own implementation, like a custom map loader or something?
Which would mean we could do this.
```
protected static int unsignedByteToInt (byte b) {
		return TiledMapLoaderUtils.unsignedByteToInt(b);
}

protected static FileHandle getRelativeFileHandle (FileHandle file, String path) {
		 return TiledMapLoaderUtils.getRelativeFileHandle(file, path);
}
```
Personally, this looks like ugly clutter for such small changes and in the libGDX codebase they are only being called by the child classes of the Tmx loader which I would update to use the new utility methods anyway. But I don't know, should we even be worried this? Am I overthinking these small changes?

Reply:
I am personally not a big fan of `utils` classes but I am also not a big fan of a lot of inheritance levels either :D But since we won't make it perfect in this PR I'd go for a base class as you initially mentioned. That way you keep the methods and people who are overriding it don't get a broken code.

Reply:
Ok cool. That probably is the way to go.
So my plan of attack would be to start with a universal parent class which both the BaseTmxMapLoader and BaseTmjMapLoader can extend and derive all their currently duplicated methods and fields from. This way we can keep all relevant map related methods together and gives us a place for a much larger refactor in the future.